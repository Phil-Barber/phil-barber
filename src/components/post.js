import React from 'react';
import styled from 'styled-components';
import { Link } from 'gatsby';
import { useHover } from '../hooks/useHover';

const PostContainer = styled.div`
  padding: ${({ theme }) => theme.spacing.xxSmall};
  padding-bottom: 0;
  transition: background-color 1s ease-out;
  text-align: left;
  :hover {
    background-color: white;
  }
`;

const Excerpt = styled.p`
  visibility: hidden;
  max-height: 0px;
  overflow: hidden;
  transition: max-height 1s, visibility 1s;
  position: relative;

  ${({ isHovered }) =>
    isHovered &&
    `
    max-height: 100px;
    visibility: visible;
  `};

  :after {
    content: '';
    width: 100%;
    height: 100%;
    position: absolute;
    left: 0;
    top: 0;
    background: linear-gradient(
      transparent 40px,
      ${({ theme }) => theme.colors.white}
    );
  }
`;

const StyledLink = styled(Link)`
  text-decoration: none;
  color: inherit;
`;

const Date = styled.span`
  color: #667269;
`;

export const Post = ({ fields, frontmatter, excerpt }) => {
  const [hoverRef, isHovered] = useHover();

  return (
    <div ref={hoverRef}>
      <PostContainer>
        <StyledLink to={fields.slug}>
          <h3>
            {frontmatter.title}
            <Date> â€” {frontmatter.dateCompleted}</Date>
          </h3>
          <Excerpt isHovered={isHovered}>{excerpt}</Excerpt>
        </StyledLink>
      </PostContainer>
    </div>
  );
};
