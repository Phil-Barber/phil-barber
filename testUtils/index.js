import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { theme } from '../src/style';
import '@testing-library/jest-dom/extend-expect';

const customRender = (ui, options, store) => {
  const AllTheProviders = ({ children }) => (
    <ThemeProvider theme={theme}>{children}</ThemeProvider>
  );
  return render(ui, { wrapper: AllTheProviders, ...options });
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
