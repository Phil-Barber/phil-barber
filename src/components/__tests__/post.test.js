import React from 'react';
import { render, fireEvent } from '../../../testUtils';
import 'jest-styled-components';
import { Post } from '../post';

describe('Post', () => {
  const defaultProps = {
    fields: {
      slug: 'slug',
    },
    frontmatter: {
      title: 'Test title',
      dateCompleted: '2020-03-01',
    },
    except: 'This is an excerpt',
  };

  let component;

  const performHover = (component) => {
    fireEvent.mouseOver(component.getByTestId('post-hover-div'));
  };

  beforeEach(() => {
    component = render(<Post {...defaultProps} />);
  });

  it('matches snapshot', () => {
    expect(component.container).toMatchSnapshot();
  });

  describe('excerpt styling', () => {
    const getExcerpt = (component) => component.getByTestId('excerpt');

    it('has max-height 0 and is hidden when not hovered', () => {
      const excerpt = getExcerpt(component);
      expect(excerpt).toHaveStyleRule('max-height', '0px');
      expect(excerpt).toHaveStyleRule('visibility', 'hidden');
    });

    it("doesn't have max-height 0 and is visible when hovered", () => {
      performHover(component);
      const excerpt = getExcerpt(component);
      expect(excerpt).not.toHaveStyleRule('max-height', '0px');
      expect(excerpt).toHaveStyleRule('visibility', 'visible');
    });
  });

  describe('PostContainer', () => {
    it('has white background on hover', () => {
      expect(
        component.getByTestId('post-container')
      ).toHaveStyleRule('background-color', 'white', { modifier: ':hover' });
    });
  });
});
