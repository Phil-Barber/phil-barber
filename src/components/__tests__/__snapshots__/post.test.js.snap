// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post matches snapshot 1`] = `
.c0 {
  padding: 4px;
  padding-bottom: 0;
  -webkit-transition: background-color 1s ease-out;
  transition: background-color 1s ease-out;
  text-align: left;
}

.c0:hover {
  background-color: white;
}

.c3 {
  visibility: hidden;
  max-height: 0px;
  overflow: hidden;
  -webkit-transition: max-height 1s,visibility 1s;
  transition: max-height 1s,visibility 1s;
  position: relative;
}

.c3:after {
  content: '';
  width: 100%;
  height: 100%;
  position: absolute;
  left: 0;
  top: 0;
  background: linear-gradient( transparent 40px, #fff );
}

.c1 {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
}

.c2 {
  color: #667269;
}

<div>
  <div
    data-testid="post-hover-div"
  >
    <div
      class="c0"
      data-testid="post-container"
    >
      <a
        class="c1"
        href="slug"
      >
        <h3>
          Test title
          <span
            class="c2"
          >
             â€” 
            2020-03-01
          </span>
        </h3>
        <p
          class="c3"
          data-testid="excerpt"
        />
      </a>
    </div>
  </div>
</div>
`;
